### ISSUE TITLE
missing raii_thread, detached_thread

### ISSUE THREAD

**Issue by cubbimew** [microsoft/GSL#471]
CppCoreGuidelines rules [CP.25: Prefer gsl::raii_thread over std::thread unless you plan to detach()](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rconc-raii_thread), [CP.30: Do not pass pointers to local variables to non-raii_threads](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rconc-pass) and [CP.26: Prefer gsl::detached_thread over std::thread if you plan to detach()](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rconc-detached_thread) mention `gsl::raii_thread` and `gsl::detached_thread` that appear to be unimplemented.

**Comment by neilmacintosh** [comment]
Thanks for pointing that out @cubbimew. It's nice to have an issue to track getting the implementation done.

They are somewhat down the implementation priority list at the moment, unless anyone wants to submit a PR.

**Comment by galik** [comment]
I happen to have these laying about that have not been tested to any real degree yet. I don't know if this is what they were intended to be. Real world template classes seem to be a whole lot more complicated looking than the ones I make...
```cpp
class detached_thread
{
public:
	template<typename Callable, typename... Args>
	explicit detached_thread(Callable&& f, Args&&... args)
	: t(std::forward<Callable>(f), std::forward<Args>(args)...) {t.detach();}

private:
	std::thread t;
};

class raii_thread
{
public:
	raii_thread(raii_thread const&) = delete;
	raii_thread(raii_thread&& other): t(std::move(other.t)) {}

	template<typename Callable, typename... Args>
	raii_thread(Callable&& f, Args&&... args)
	: t(std::forward<Callable>(f), std::forward<Args>(args)...) {}

	~raii_thread() { if(t.joinable()) t.join(); }

private:
	std::thread t;
};
```
If there is more to it than this then you probably want to have  *real* expert provide something more appropriate. 😄

**Comment by neilmacintosh** [comment]
@galik I appreciate any starting point ;-)

Would it be possible to throw together a PR containing these? That way they pass through the CLA process, so we get the legal clearance to use the code (not that I'm accusing you of trying to do anything shady, just trying to dot my i's and cross my t's). Having them as a PR would also make it easier for people to review\contribute to their evolution as necessary.

**Comment by galik** [comment]
@neilmacintosh Sure. I just posted what code I currently had here to make sure I had the functionality correct. Regarding `gsl::detached_thread` it doesn't seem to make much sense to implement moving,  copying or joining. In fact the sooner the object dies the better as far as I can see. Would I delete the irrelevant functions or keep them to mimic equivalence with `std::thread`?

**Comment by JordanMaples** [comment]
Maintainers' call: We never got around to implementing gsl::joining_thread, now that  C++20 has std::jthread, please use that instead. Sorry this took so long.
